### PR TITLE
revert resource disablement, 1.10

### PR DESCRIPTION
**Pull request by deads2k**
This is not an automated pick!

An attempt to revert for https://github.com/kubernetes/kubernetes/issues/62044.  It is not clean.

```release-note
Restores ability to enable/disable resource within an API GroupVersion individually.
```

### CODE REVIEW

**Comment by deads2k** [comment]
/assign @lavalamp 
/assign @sttts 

We could merge this for 1.10 and pursue a cleaner fix for 1.11.  That's probably the most palatable option we have.

**Comment by lavalamp** [comment]
I think I'm OK with this as long as it doesn't get dropped for 1.11.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -23,24 +23,37 @@ import (
...
 	informers "k8s.io/kube-aggregator/pkg/client/informers/internalversion"
 	listers "k8s.io/kube-aggregator/pkg/client/listers/apiregistration/internalversion"
 	openapicontroller "k8s.io/kube-aggregator/pkg/controllers/openapi"
 	statuscontrollers "k8s.io/kube-aggregator/pkg/controllers/status"
-	apiservicerest "k8s.io/kube-aggregator/pkg/registry/apiservice/rest"
+	apiservicestorage "k8s.io/kube-aggregator/pkg/registry/apiservice/etcd"
+)
+
+var (
+	groupFactoryRegistry = make(announced.APIGroupFactoryRegistry)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: missing the corresponding (file) removals to the aggregator changes.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -57,6 +57,7 @@ func init() {
 
 	// we need to add the options (like ListOptions) to empty v1
 	metav1.AddToGroupVersion(aggregatorscheme.Scheme, schema.GroupVersion{Group: "", Version: "v1"})
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Group: "", Version: "v1"})
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: interesting that this is related to the PR.

**Comment by sttts** [comment]
@deads2k which of the commits were unclean during revert?

**Comment by deads2k** [comment]
> @deads2k which of the commits were unclean during revert?

All of them

**Comment by lavalamp** [comment]
I would prefer to have this ASAP.

**Comment by MaciekPytel** [comment]
In that case can someone from sig-apimachinery review this today (assuming it's finished)? Also can we squash merge commit from fork?

@deads2k @sttts @lavalamp

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: fixed

**Comment by sttts** [comment]
Lgtm. Due to its size I would prefer having some more eyes on it @caesarxuchao @hzxuzhonghu

**Comment by deads2k** [comment]
> Lgtm. Due to its size I would prefer having some more eyes on it @caesarxuchao @hzxuzhonghu



/assign @caesarxuchao 
/assign @hzxuzhonghu

**Comment by hzxuzhonghu** [comment]
LGTM except some nits

File: pkg/registry/policy/rest/storage_policy.go [hunk 3]
```diff
@@ -35,22 +35,27 @@ func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorag
...
 
-	storage["podsecuritypolicies"] = pspstore.NewREST(restOptionsGetter)
+	// poddisruptionbudgets
+	if apiResourceConfigSource.ResourceEnabled(version.WithResource("poddisruptionbudgets")) {
+		poddisruptionbudgetStorage, poddisruptionbudgetStatusStorage := poddisruptionbudgetstore.NewREST(restOptionsGetter)
+		storage["poddisruptionbudgets"] = poddisruptionbudgetStorage
+		storage["poddisruptionbudgets/status"] = poddisruptionbudgetStatusStorage
+	}
 
+	if apiResourceConfigSource.ResourceEnabled(version.WithResource("podsecuritypolicies")) {
```
Comment (hzxuzhonghu) [inline pkg/registry/policy/rest/storage_policy.go]: Should here also support `podsecuritypolicy` as extentions group?

Comment (deads2k) [reply pkg/registry/policy/rest/storage_policy.go]: > Should here also support podsecuritypolicy as extentions group?

No.  separate makes sense

**Comment by deads2k** [comment]
nits addressed

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -730,6 +730,19 @@ func defaultOptions(s *options.ServerRunOptions) error {
 		}
 	}
 
+	// TODO: remove when we stop supporting the legacy group version.
+	if s.APIEnablement.RuntimeConfig != nil {
+		for key, value := range s.APIEnablement.RuntimeConfig {
+			if key == "v1" || strings.HasPrefix(key, "v1/") ||
+				key == "api/v1" || strings.HasPrefix(key, "api/v1/") {
```
Comment (hzxuzhonghu) [inline cmd/kube-apiserver/app/server.go]: This can not enable disable resources in core v1

Comment (hzxuzhonghu) [reply cmd/kube-apiserver/app/server.go]: see https://github.com/kubernetes/kubernetes/pull/62254/commits/2a9ebb3f6678fe70348f23fa40f5c6b234078a3e#diff-3d1aa7e5d43db965581b230647a90de2R71

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 2 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: Schema here duplicated with aggregatorscheme.Scheme. I think keep one is ok.

**Comment by MaciekPytel** [comment]
This has been reviewed by a lot of people in sig-apimachinery, so I'm adding cherrypick-approved. Someone still needs to step forward and a final lgtm. @sttts @hzxuzhonghu @lavalamp 

@deads2k Can you rebase?

**Comment by lavalamp** [comment]
I'll LGTM if we rebase. (Or is the reimplementation PR ready? If that's
ready we could take it instead.)

**Comment by deads2k** [comment]
> I'll LGTM if we rebase. (Or is the reimplementation PR ready? If that's
> ready we could take it instead.)

it's rebased.

**Comment by mikedanese** [comment]
This appears have broken GKE test on release-1.10 branch (at least more than it was broken before :) ):

https://k8s-testgrid.appspot.com/sig-release-1.10-blocking#gci-gke-1.10

Before this change, masters were coming up. After this change, the master isn't coming up and the apiserver is crash looping with:

```
error in initializing storage factory: group version apiregistration.k8s.io/v1beta1 that has not been registered
```

**Comment by MaciekPytel** [comment]
@deads2k @lavalamp I can't cut the release with all GKE release-blocking suites failing due to masters not coming up. Is there some quick fix we can apply here? Otherwise I think the safest option would be to rollback this PR for 1.10.2 and try again after the release is cut, so there is enough time to fix any issues.

**Comment by deads2k** [comment]
> Otherwise I think the safest option would be to rollback this PR for 1.10.2

ok

**Comment by lavalamp** [comment]
I wanted to get someone to debug this but my afternoon was pretty packed :(

**Comment by MaciekPytel** [comment]
I'm reverting this to unblock 1.10.2. We can try fixed version after the release if you want.

**Comment by perotinus** [comment]
I did a bit of testing with this PR at HEAD before it was reverted, and found that disabling `api/v1` resources didn't work. I was able to individually disable resources in the `batch` group, but could not individually enable/disable resources in the `api/v1` group. We need to be able to disable/enable individual resources in `api/v1` specifically.

**Comment by mbohlool** [comment]
Should we revert this on master too? GKE tests are still failing on master. @deads2k @lavalamp

**Pull request merged**